docs/rfc/rfc-1.adoc

```diff
@@ -435,13 +435,64 @@ As time passes and operators drop out of the system, a wallet becomes at risk
 of being able to meet the 51-of-100 threshold to produce signatures.
 Additionally, we want to avoid situations where operators are the custodians of
 a wallet for extended periods. To avoid these issues, once a wallet is older
-than the `wallet_max_age`, or if it drops below the liveness threshold (say, below 70 on
-a <<heartbeat,heartbeat>>), we motion to transfer the funds to another randomly
-selected wallet.
+than the `wallet_max_age`, or if it drops below the liveness threshold (say,
+below 70 on a <<heartbeat,heartbeat>>), we motion to
+<<closing-a-wallet,transfer the funds>> to another randomly selected wallet.
 
 Once a wallet no longer has funds and is not the primary wallet for new
-deposits, it can be closed and operators are no longer required to maintain
-it.
+deposits, it can be <<closing-a-wallet,closed>> and operators are no longer
+required to maintain it.
+
+[[closing-a-wallet]]
+==== Closing A Wallet
+
+Governable Parameters:
+
+- `heartbeat`: The number of group members required for a
+  <<heartbeat,heartbeat>> to be successful.
+- `random_transfer_depth`: The depth (sorted youngest to oldest) at which we
+  randomly select a new wallet to be the custodian of the old wallet's funds
+  when closing a wallet. For instance, a depth of 3 means that the 3 youngest
+  wallets can be randomly chosen.
+- `wallet_closure_timeout`: The amount of time that a wallet has to
+  successfully move its funds to another wallet and inform the ethereum chain
+  before it is at risk of punishment.
+- `wallet_max_age`: The oldest we allow a wallet to become before we transfer the funds
+  to a randomly selected wallet.
+- `wallet_min_btc`: The smallest amount of btc a wallet can hold after a
+  redemption before we attempt to close the wallet and transfer the funds to a
+  randomly selected wallet.
+
+When a wallet drops below the `heartbeat` threshold during a
+<<heartbeat,heartbeat>> or ends a <<redemption,redemption>> with less than
+`wallet_min_btc` remaining, or exceeds the `wallet_max_age`, the operators
+begin the process of closing it. If there are still funds remaining, the
+<<operator-only,first operator>> posts a <<operator-only,reimbursable>>
+transaction to the ethereum chain declaring an intention to close the wallet.
+
+That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and the
+operators use the hash of the previous intention-to-close transaction as a
+source of randomness to choose viable (it must also have a passing heartbeat
+and be above `wallet_min_btc`, and be within the `random_transfer_depth`)
+wallet to transfer the remaining funds to. They construct a P2PKH transaction
+moving the wallet's main UTXO to that wallet.
```

Comment:
We [can not](https://docs.soliditylang.org/en/latest/units-and-global-variables.html) have the hash of the previous intention-to-close transaction on-chain so we will not be able to validate on-chain if the funds were moved to the right wallet. If we do not have to validate it on-chain and we can trust the majority in the wallet to move the funds to any of the `random_transfer_depth` new wallets, this is fine. But if we want to validate it on-chain, then we need to choose something else as the seed.

Options that come to my mind:
- ask the random beacon
- use the most recently created wallet's pubkey hash
- use relay entry from the latest DKG

All these options have their pros and cons. Random beacon is probably the most secure one but it adds some complexity given that the beacon can be busy at the moment of failing the heartbeat and we do want to start moving funds immediately.

On the first sight, hash of the pubkey of the most recently created wallet does not sound that bad but worth thinking more about it cc @eth-r

Reply:
The transaction hash is not available from Solidity function, see the parameters [here](https://docs.soliditylang.org/en/latest/units-and-global-variables.html). We can have it off-chain but not on-chain.

Reply:
I think it's fine that we just have it off-chain, since we're just using the information as a way to short-cut consensus.

Here's how I imagine this plays out in the happy path:

+ We detect that we need to transfer funds, and that gets posted to ethereum with a particular hash
+ The wallet reads that message, uses that hash off-chain to all seed the same RNG and pick a uniform random number between [0, `random_transfer_depth`], and picks a wallet from that number, and transfers the funds to that wallet
+ No one challenges the tx because it isn't fraudulent

Unchallengable Bad Path

+ We detect that we need to transfer funds, and that gets posted to ethereum with a particular hash
+ The wallet reads that message, colludes to send it to a wallet within [0, `random_transfer_depth`], and transfers the funds to that wallet
+ No one challenges the tx because it doesn't look fraudulent

(this requires a 51-of-100 attack with modified clients)

Challengable Bad Path

+ We detect that we need to transfer funds, and that gets posted to ethereum with a particular hash
+ The wallet reads that message, colludes to send it to a wallet outside of [0, `random_transfer_depth`], and transfers the funds to that wallet (could be a totally separate wallet)
+ challenge/response sighash etc fraud-proof process

(this requires a 51-of-100 attack with modified clients)

---

So, we open ourselves up to a 51-of-100 attack where the wallet is able to move the funds to a wallet *that they already control* to another wallet of their choosing within `[0, random_transfer_depth]` and have it not seem like fraud.

---

Asking the random beacon sounds gas expensive, using the wallet pubkey hash is past, static information which means that all wallet closures would go to the same wallet until a new wallet is created unless you also had some sort of indexing, and then folks could manipulate the index to get the funds to move into the wallet they wanted.

Can you elaborate on what the relay entry is?

So my main question at the moment is: how expensive is the attack I'm describing above vs paying the random beacon to prevent that attack? There's a decent chance that we're overengineering this at the moment and we could just do the off-chain consensus based on the intention-to-close transaction and upgrade to random-beacon later after more analysis later